### ISSUE TITLE
SAL inference kernel roadmap: prefill, decode, KV cache, LoRA e quantização

### ISSUE THREAD

**Issue by v1olegrace** [v1olegrace/triton-kernel-lab#2]
# Objetivo

Transformar o `triton-kernel-lab` em uma bancada de kernels que possa ser usada diretamente como fast path de inferência para o stack SAL: contexto longo, prefill grande, decode autoregressivo longo, adapters LoRA e eventual quantização 4/8-bit.

A regra desta trilha é: **nenhum kernel conta como pronto só por vencer um microbenchmark**. Cada kernel precisa passar por `correctness -> microbenchmark -> roofline -> NCU -> SAL end-to-end`.

## Baseline já existente no laboratório

O repo já possui implementações/documentação de:

- [x] GEMM / matmul
- [x] fused softmax
- [x] RMSNorm
- [x] residual + RMSNorm
- [x] RoPE
- [x] SwiGLU
- [x] FlashAttention

Portanto o foco SAL deve ser ampliar o laboratório para **serving de LLM real**, sobretudo decode e memória.

---

# 1. Contrato de benchmark SAL

Todo kernel novo ou otimizado deve registrar, além do benchmark tradicional:

- [ ] `latency_us` p50/p95
- [ ] `effective_GBps`
- [ ] `effective_TFLOPs`
- [ ] fração do roof empírico da GPU
- [ ] bytes lidos/escritos estimados
- [ ] occupancy
- [ ] registers/thread
- [ ] shared memory / CTA
- [ ] número de launches por bloco transformer
- [ ] comparação PyTorch eager
- [ ] comparação PyTorch compile quando aplicável
- [ ] comparação Triton baseline
- [ ] comparação vendor baseline quando existir

## Métricas end-to-end obrigatórias

- [ ] TTFT — Time To First Token
- [ ] TPOT — Time Per Output Token
- [ ] tokens/s decode
- [ ] tokens/s prefill
- [ ] peak VRAM
- [ ] KV-cache bytes/token
- [ ] throughput por batch
- [ ] throughput por sequence length

### Shape grid mínima

Testar separadamente prefill e decode.

**Prefill sequence lengths:**

- [ ] 512
- [ ] 2K
- [ ] 4K
- [ ] 8K
- [ ] 16K
- [ ] 32K

**Decode KV lengths:**

- [ ] 512
- [ ] 2K
- [ ] 4K
- [ ] 8K
- [ ] 16K
- [ ] 32K

**Batch:**

- [ ] 1
- [ ] 2
- [ ] 4
- [ ] 8

**Dtypes prioritários:**

- [ ] FP32 reference apenas para correctness
- [ ] FP16
- [ ] BF16 onde suportado
- [ ] INT8/FP8 em trilha posterior
- [ ] INT4 weight-only em trilha de quantização

---

# 2. GEMM — adaptar para shapes reais de LLM

Arquivo atual de referência: `src/tklab/kernels/matmul.py`.

Não basta medir M=N=K quadrado. Adicionar shapes de projeções reais:

- [ ] Q projection: `[B*T, H] x [H, Hq]`
- [ ] K projection: `[B*T, H] x [H, Hkv]`
- [ ] V projection: `[B*T, H] x [H, Hkv]`
- [ ] O projection: `[B*T, H] x [H, H]`
- [ ] MLP gate/up: `[B*T, H] x [H, I]`
- [ ] MLP down: `[B*T, I] x [I, H]`

### Shapes modelo-alvo

Criar profiles para pelo menos:

- [ ] 8B-class
- [ ] 12–14B-class
- [ ] 32B-class

Cada profile deve carregar `hidden_size`, `intermediate_size`, `num_q_heads`, `num_kv_heads`, `head_dim`.

### Acceptance

- [ ] erro numérico compatível com PyTorch/cublas para dtype escolhido
- [ ] pelo menos 90% do roof empírico em shapes compute-bound representativos ou justificar gargalo
- [ ] regressão automática para shapes skinny/tall típicos de decode

---

# 3. RMSNorm / residual RMSNorm

Arquivos atuais:

- `src/tklab/kernels/rms_norm.py`
- `src/tklab/kernels/residual_rms_norm.py`

Especificações SAL:

- [ ] benchmark batch-token flatten `[B*T, H]`
- [ ] benchmark decode `[B, H]`
- [ ] fused residual add + RMSNorm
- [ ] opcional: fused residual + RMSNorm + cast
- [ ] medir bandwidth efetiva, não TFLOPs como métrica principal
- [ ] testar H dos model profiles 8B/14B/32B

### Acceptance

- [ ] memory-bound próximo do roof de bandwidth empírico
- [ ] nenhuma materialização intermediária desnecessária
- [ ] erro máx / erro relativo registrados por dtype

---

# 4. SwiGLU / MLP fusion

Arquivo atual: `src/tklab/kernels/swiglu.py`.

Expandir para caminho de transformer real:

- [ ] gate activation `SiLU(gate)`
- [ ] multiply `SiLU(gate) * up`
- [ ] evitar intermediários quando possível
- [ ] benchmark junto do custo dos GEMMs adjacentes
- [ ] estudar fusão parcial gate/up -> SwiGLU

### Acceptance

- [ ] demonstrar redução de HBM traffic
- [ ] mostrar ganho end-to-end no bloco MLP, não apenas no kernel isolado

---

# 5. RoPE para contexto longo

Arquivo atual: `src/tklab/kernels/rope.py`.

Especificações:

- [ ] Q e K separados
- [ ] Q+K no mesmo launch quando vantajoso
- [ ] support offset de posição para decode incremental
- [ ] posições > 32K
- [ ] cos/sin precomputed cache
- [ ] comparar cache lookup vs compute-on-the-fly
- [ ] layout compatível com GQA

### Acceptance

- [ ] equivalência com implementação de referência
- [ ] custo de RoPE por token medido no decode
- [ ] sem reprocessar prefixo durante decode

---

# 6. FlashAttention: separar PREFILL de DECODE

Arquivo atual: `src/tklab/kernels/flash_attention.py`.

A implementação de prefill não deve ser tratada como decode kernel.

## Prefill

- [ ] causal attention
- [ ] GQA/MQA
- [ ] sequence lengths até 32K no benchmark local
- [ ] online softmax numericamente estável
- [ ] máscara causal sem materialização
- [ ] comparar com SDPA / FlashAttention disponível no ambiente

## Decode

Criar kernel específico para `Q_len = 1` ou pequeno chunk de queries:

- [ ] paged/contiguous KV read baseline
- [ ] GQA
- [ ] head_dim 64/128 prioritário
- [ ] KV length até 32K+
- [ ] minimizar launch overhead
- [ ] reduzir leituras redundantes de K/V

### Acceptance

Prefill:
- [ ] throughput em tokens/s e TFLOPs

Decode:
- [ ] TPOT e GB/s são métricas primárias
- [ ] comparar contra SDPA/vendor path

---

# 7. KV cache — PRIORIDADE MÁXIMA para SAL

Adicionar módulos novos:

- [ ] `kv_cache_append.py`
- [ ] `kv_cache_gather.py`
- [ ] `paged_kv_cache.py` ou baseline equivalente

## Append

Contrato aproximado:

```text
K_new, V_new, slot_mapping -> KV_cache
```

- [ ] sem cópia do prefixo
- [ ] append por token
- [ ] append por chunk
- [ ] GQA/MQA layouts
- [ ] coalesced writes

## Gather/read

- [ ] contiguous baseline
- [ ] paged layout
- [ ] block table / slot mapping
- [ ] gather de K/V para decode attention sem materializar tensor intermediário quando possível

## KV quantization — fase posterior

- [ ] FP16/BF16 baseline
- [ ] FP8 ou INT8 KV
- [ ] medir degradação numérica separadamente de ganho de VRAM

### Métrica central

Registrar:

```text
KV_bytes_per_token
KV_read_GB_per_generated_token
```

### Acceptance

- [ ] custo por token cresce conforme esperado com context length
- [ ] nenhum full-cache copy em decode
- [ ] benchmark B=1 e continuous-batching-like shapes

---

# 8. LoRA fast path

Adicionar:

- [ ] `lora_matmul.py`
- [ ] benchmark multi-adapter

Operação alvo:

```text
y = Wx + scale * B(Ax)
```

com `scale = alpha / r`.

Ranks prioritários:

- [ ] r=8
- [ ] r=16
- [ ] r=32
- [ ] r=64

### Experimentos

- [ ] baseline em três GEMMs/ops separados
- [ ] reduzir materialização de `Ax`
- [ ] fused accumulate na saída base quando possível
- [ ] adapter batching
- [ ] medir custo de trocar adapter

### Caso SAL

Avaliar fast path para adapters de:

- [ ] world/style
- [ ] dialogue
- [ ] character-specific

Não assumir que separar adapters melhora qualidade; esta parte é apenas infraestrutura para permitir a ablation.

### Acceptance

- [ ] numericamente equivalente a PEFT/reference
- [ ] overhead LoRA explícito em µs/token
- [ ] medir ganho real em TTFT/TPOT

---

# 9. Quantized matmul

Criar trilha dedicada após FP16/BF16 estar sólida.

- [ ] W8A16 baseline
- [ ] W4A16 weight-only
- [ ] unpack/dequant fused com matmul quando possível
- [ ] per-group scales
- [ ] group sizes 32/64/128
- [ ] zero-point opcional conforme formato
- [ ] layout/padding documentados

Comparar contra implementações existentes no ambiente quando possível.

### Métricas

- [ ] model weight bytes
- [ ] GB/s
- [ ] tokens/s
- [ ] dequant overhead
- [ ] erro de saída vs FP16

### Acceptance

Um kernel quantizado só é vitória se melhorar a aplicação end-to-end; reduzir bytes sem melhorar TPOT/throughput não conta como sucesso.

---

# 10. Fused QKV projection

Depois de GEMM e RoPE estáveis:

- [ ] investigar Q/K/V em um caminho de projeção único ou concatenado quando o layout do modelo permitir
- [ ] aplicar RoPE em Q/K sem round-trip adicional à HBM quando possível
- [ ] escrever K/V diretamente no cache

Target conceitual:

```text
x
 -> QKV projection
 -> RoPE(Q,K)
 -> K/V cache append
 -> Q para attention
```

Evitar intermediários é mais importante que simplesmente reduzir número nominal de kernels.

---

# 11. SAL end-to-end workload

Adicionar benchmark que represente a aplicação real.

## Workload A — short scene

```text
prompt: 4K tokens
generation: 1K tokens
batch: 1
```

## Workload B — chapter

```text
prompt: 16K tokens
generation: 4K tokens
batch: 1
```

## Workload C — long-context chapter

```text
prompt: 32K tokens
generation: 4K tokens
batch: 1
```

## Workload D — small serving batch

```text
prompt: 4–16K variable
generation: 512–2K variable
batch/concurrent sequences: 4–8
```

Registrar:

- [ ] TTFT
- [ ] TPOT
- [ ] prefill tokens/s
- [ ] decode tokens/s
- [ ] peak VRAM
- [ ] KV VRAM
- [ ] total kernel launches/token
- [ ] % tempo GEMM
- [ ] % tempo attention
- [ ] % tempo KV movement
- [ ] % tempo normalization/activation

---

# 12. Critério de prioridade

## P0 — fazer primeiro

- [ ] model-shape GEMM benchmark
- [ ] decode-specific attention benchmark
- [ ] KV cache append/read
- [ ] SAL TTFT/TPOT harness

## P1

- [ ] LoRA matmul
- [ ] long-context RoPE decode path
- [ ] fused residual RMSNorm tuning
- [ ] SwiGLU/MLP memory-traffic tuning

## P2

- [ ] W8A16
- [ ] W4A16
- [ ] KV quantization
- [ ] fused QKV + RoPE + cache write experiments

## P3 — só depois de evidência do profiler

- [ ] reescrever atenção completa se vendor/FlashAttention já dominar o workload
- [ ] fusões muito grandes que elevem register pressure/occupancy loss
- [ ] customizações específicas de SAL sem ganho generalizável

---

# 13. Regra para merge de kernel SAL

Um kernel desta trilha só deve ser marcado `production_candidate` se:

- [ ] correctness passa em seed fixa + randomized shapes
- [ ] edge cases passam
- [ ] benchmark é reproduzível
- [ ] resultado bruto salvo em artefato versionado
- [ ] roofline classificado corretamente
- [ ] profiler confirma a hipótese de otimização
- [ ] existe comparação com baseline forte
- [ ] não há regressão relevante em shapes adjacentes
- [ ] há impacto mensurável em TTFT, TPOT ou throughput SAL

---

# 14. Ordem recomendada de implementação

```text
model-shape GEMM
  -> KV cache append/read
  -> decode attention
  -> SAL TTFT/TPOT harness
  -> LoRA matmul
  -> RoPE incremental tuning
  -> residual RMSNorm / SwiGLU block tuning
  -> quantized matmul
  -> KV quantization
  -> fused QKV/RoPE/cache experiments
```

O objetivo final não é substituir vLLM/SGLang/TensorRT-LLM inteiro. É construir um laboratório capaz de provar, kernel por kernel, **onde uma otimização própria realmente melhora um workload LLM de contexto longo**, e depois usar esses componentes no ecossistema SAL.
